File: clients/Cpp/DebugInterface.cpp
#include "DebugInterface.hpp"
#include "model/ClientMessage.hpp"

DebugInterface::DebugInterface(const std::shared_ptr<InputStream>& inputStream, const std::shared_ptr<OutputStream>& outputStream)
    : inputStream(inputStream)
    , outputStream(outputStream)
{
}

void DebugInterface::send(const DebugCommand& command)
{
    // TODO: Construct actual message, this is a hack :)
    outputStream->write(ClientMessage::DebugMessage::TAG);
    command.writeTo(*outputStream);
    outputStream->flush();
}

DebugState DebugInterface::getState()
{
    ClientMessage::RequestDebugState().writeTo(*outputStream);
    outputStream->flush();
    return DebugState::readFrom(*inputStream);
}

File: clients/Scala/src/main/scala/DebugInterface.scala
import java.io.InputStream
import java.io.OutputStream

class DebugInterface(private val inputStream: InputStream, private val outputStream: OutputStream) {
  def send(command: model.DebugCommand) {
    model.ClientMessage.DebugMessage(command).writeTo(outputStream)
    outputStream.flush()
  }
  def getState(): model.DebugState = {
    model.ClientMessage.RequestDebugState().writeTo(outputStream)
    outputStream.flush()
    model.DebugState.readFrom(inputStream)
  }
}